token-swap/program/src/processor.rs

```diff
@@ -149,6 +153,43 @@ impl Processor {
             signers,
         )
     }
+    fn get_bit_at(input: u8, n: usize) -> Result<bool, ()> {
+        if n < 8 {
+            Ok(input & (1 << n) != 0)
+        } else {
+            Err(())
+        }
+    }
+
+    fn check_allowed_to_freeze(
+        token_swap: &dyn SwapState,
+        freeze_authority_info: &AccountInfo,
+    ) -> ProgramResult {
+        match token_swap.freeze_authority() {
+            COption::Some(key) => {
+                if key == *freeze_authority_info.key && freeze_authority_info.is_signer {
+                    Ok(())
+                } else {
+                    Err(SwapError::UnauthorizedToFreeze.into())
+                }
+            }
+            COption::None => Err(SwapError::UnauthorizedToFreeze.into()),
+        }
+    }
+
+    fn check_allowed_to_use(token_swap: &dyn SwapState, bit_position: usize) -> ProgramResult {
+        let bitmask: u8 = token_swap.freeze_authority_bit_mask();
+        match Self::get_bit_at(bitmask, bit_position) {
+            Ok(frozen) => {
+                if frozen {
```

Comment:
A newish best practice is to have a `msg!` explaining what's happening before throwing an error, which makes debugging from the explorer much easier.  Swap isn't doing this everywhere of course, but this new code should :smile: